### PR TITLE
support aiter backend for mimo models

### PR DESCRIPTION
**Pull request by yucshen**
## Motivation

[Xiaomi MiMo-V2-Flash](https://huggingface.co/XiaomiMiMo/MiMo-V2-Flash) is a hybrid MoE reasoning model with an unusual attention geometry: the Q/K head dimension is 192 while the V head dimension is 128 (\`qk_head_dim != v_head_dim\`). All 48 layers also use a 128-token sliding-window attention (SWA).

The SGLang aiter backend (\`AiterAttnBackend\`) previously assumed that K and V share the same head dimension throughout. This assumption is embedded in three places:

1. **Prefill** – \`mha_batch_prefill_func\` (AIter's batch-prefill kernel) internally views the K and V caches with a single \`head_dim\`, producing corrupted output when the two dims differ.
2. **Decode** – \`unified_attention\` (AIter's paged-attention kernel) has the same equal-dim assumption, causing silent memory corruption during token generation.
3. **KV-cache write** – \`launch_reshape_and_cache_flash\` uses \`key.shape[2]\` as the head dim for *both* K and V cache writes, writing V tokens into the wrong memory locations whenever \`v_head_dim < qk_head_dim\`.

Without this fix, launching MiMo-V2-Flash with \`--attention-backend aiter\` produces incorrect output or crashes. This PR adds targeted fallback paths that activate only when \`layer.qk_head_dim != layer.v_head_dim\`, leaving the existing fast paths completely unchanged for standard square-attention models.

## Modifications

All changes are confined to \`python/sglang/srt/layers/attention/aiter_backend.py\`.

### 1. Prefill fallback (\`forward_extend\`)

When \`layer.qk_head_dim != layer.v_head_dim\`, skip \`mha_batch_prefill_func\` and call the Triton \`extend_attention_fwd\` kernel instead. This kernel accepts arbitrary \`Lk\` and \`Lv\` values. The prefix KV indices are rebuilt from \`extend_prefix_lens\` so that only the cached (prefix) tokens are passed as context, exactly as required by the kernel's API.

### 2. Decode fallback (\`forward_decode\`)

When \`layer.qk_head_dim != layer.v_head_dim\`, skip \`unified_attention\` and call the Triton \`decode_attention_fwd\` kernel. This kernel accepts separate key and value buffers with independent head dims. Key implementation details:

- **Flat index arrays** – \`decode_attention_fwd\` expects a flat per-token index array (not paged). A new helper \`_build_nonsq_flat_indices\` builds these from \`req_to_token\` using the existing \`create_flashinfer_kv_indices_triton\` Triton kernel. For SWA layers, the index array is windowed to the last \`sliding_window_size\` tokens.
- **Pre-allocated buffers** – All scratch buffers (\`attn_logits\`, \`attn_lse\`, \`num_kv_splits\`, flat index arrays) are allocated once on the first call and reused, keeping the path CUDA-graph-safe (no \`torch.empty\` inside graph capture).
- **Fixed-stride buffers** – The \`kv_indices_ptr_stride\` argument to the Triton kernel is a \`tl.constexpr\`. On AMD/ROCm, changing a constexpr value between calls triggers a Triton recompile that crashes. Two pre-allocated fixed-shape index buffers with stride equal to \`max_context_len\` are used so the constexpr never changes.
- **FP8 KV dequantization** – \`decode_attention_fwd\` does not support FP8 KV input. When \`kv_cache_dtype == fp8\`, the K and V buffers are dequantized to BF16 before the call using pre-allocated BF16 scratch tensors.
- **Scale extraction** – \`k_scale.item()\` would be a CPU-GPU sync inside the forward pass. Scales are extracted once during metadata initialization and cached as Python floats.

### 3. KV-cache write fix

When \`qk_head_dim != v_head_dim\`, bypass \`launch_reshape_and_cache_flash\` (which uses a single head dim for both K and V) and instead write directly:

\`\`\`python
k_cache[loc] = k.view(-1, layer.tp_k_head_num, layer.qk_head_dim)
v_cache[loc] = v.view(-1, layer.tp_v_head_num, layer.v_head_dim)
\`\`\`

For SWA layers, \`loc\` is obtained from \`pool.full_to_swa_index_mapping[cache_loc]\`. This applies to both the extend (prefill) and decode cache-write paths.

## Accuracy Tests

**Model:** XiaomiMiMo/MiMo-V2-Flash
**Hardware:** 4x AMD Instinct MI355X (TP=4)
**Backend:** \`--attention-backend aiter\`
**Eval:** GSM8K 8-shot completion, 1319 questions

| Metric | Score |
| ------ | ----- |
| Accuracy (strict match) | **79.9%** (1054/1319) |
| Invalid responses | 0.0% |
| Output throughput | 2013 tok/s |

## Speed Tests and Profiling

**Hardware:** 4x AMD Instinct MI355X (TP=4)
**Backend:** \`--attention-backend aiter\`
**Tool:** \`python3 -m sglang.bench_serving --backend sglang --dataset-name random\`

| Input len | Output len | Concurrency | Input tok/s | Output tok/s | Mean TTFT (ms) | Mean ITL (ms) |
| --------- | ---------- | ----------- | ----------- | ------------ | -------------- | ------------- |
| 256 | 128 | 1 | 414 | 141 | 50.1 | 6.2 |
| 1024 | 512 | 16 | 2,379 | 1,077 | 107.6 | 12.9 |
| 4096 | 512 | 16 | **8,343** | 958 | 233.1 | 14.3 |

At concurrency=16 with 4096-token inputs, input token throughput reaches **8,343 tok/s**.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include \`/tag-and-rerun-ci\`, \`/tag-run-ci-label\`, \`/rerun-failed-ci\`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

**Comment by sammysun0711** [comment]
Hi @yucshen, many thanks for your great contribution to support `aiter attention backend` for https://huggingface.co/XiaomiMiMo/MiMo-V2-Flash on ROCm.
Could you please kindly share following information if possible: 
- Performance gain with default `--attention-backend triton` vs `--attention-backend aiter` for 256 in/128 out, 1024/512 out, 4096 in/512 out case.
- Does `aiter attention backend` also support MTP enabled as follow? 
`--speculative-algorithm EAGLE  --speculative-num-steps 3  --speculative-eagle-topk 1  --speculative-num-draft-tokens 4 `

**Comment by yucshen** [comment]
Hi @sammysun0711, thanks for the reply.

> * Performance gain with default `--attention-backend triton` vs `--attention-backend aiter` for 256 in/128 out, 1024/512 out, 4096 in/512 out case.

This can be done. I will test it.

> * Does `aiter attention backend` also support MTP enabled as follow?
>   `--speculative-algorithm EAGLE  --speculative-num-steps 3  --speculative-eagle-topk 1  --speculative-num-draft-tokens 4 `

Short answer is no, not yet. Currently, I am working on the MTP support. Do you thinks MTP support needs to be shipped with this PR?

**Comment by sammysun0711** [comment]
> Hi @sammysun0711, thanks for the reply.
> 
> > * Performance gain with default `--attention-backend triton` vs `--attention-backend aiter` for 256 in/128 out, 1024/512 out, 4096 in/512 out case.
> 
> This can be done. I will test it.
> 
> > * Does `aiter attention backend` also support MTP enabled as follow?
> >   `--speculative-algorithm EAGLE  --speculative-num-steps 3  --speculative-eagle-topk 1  --speculative-num-draft-tokens 4 `
> 
> Short answer is no, not yet. Currently, I am working on the MTP support. Do you thinks MTP support needs to be shipped with this PR?

Hi @yucshen, thanks for your feedback. It will be great to have performance comparison data as reference. 

For MTP feature, It will be great if we can support in aiter attention backend, it will boost output token throughput especially in prefill/decode disaggregated serving use case. However, it may takes some additional effort to support MTP feature, so I suggest you may submit a follow up PR to address this issue.
